lib/galaxy/job_execution/output_collect.py

```diff
@@ -378,8 +378,17 @@ def collect_primary_datasets(job_context, output, input_ext):
                 # Before I guess pop() would just have thrown an IndexError
                 raise Exception("Problem parsing metadata fields for file %s" % filename)
             designation = fields_match.designation
+            ext = fields_match.ext
+            if ext == "input":
+                ext = input_ext
+            dbkey = fields_match.dbkey
+            if dbkey == INPUT_DBKEY_TOKEN:
+                dbkey = job_context.input_dbkey
             if filename_index == 0 and extra_file_collector.assign_primary_output and output_index == 0:
                 new_outdata_name = fields_match.name or "%s (%s)" % (outdata.name, designation)
+                outdata.change_datatype(ext)
```

Comment:
change_datatype does a couple of things we don't need to do when the dataset is new (and hat would be skipped after checking with the object store if the file exists).
```suggestion
                outdata.extension = ext
```
Setting up the metadata attributes should be handled on line 443.